src/utils.rs

```diff
@@ -0,0 +1,155 @@
+//! Utility functions for the Sprocket command line tool.
+
+use std::path::PathBuf;
+use std::fs;
+
+use anyhow::{Result, Context, bail};
+use tracing::debug;
+
+/// Represents the format of an input file.
+#[derive(Debug, PartialEq)]
+pub enum InputFormat {
+    /// JSON format
+    Json,
+    /// YAML format
+    Yaml,
+}
+
+/// Determines the format of an input file by attempting to parse it.
+pub fn detect_input_format(file_path: &PathBuf) -> Result<InputFormat> {
+    let content = fs::read_to_string(file_path)
+        .with_context(|| format!("Failed to read file: {}", file_path.display()))?;
+    
+    // Try parsing as JSON first
+    if let Ok(_) = serde_json::from_str::<serde_json::Value>(&content) {
+        debug!("File detected as JSON: {}", file_path.display());
+        return Ok(InputFormat::Json);
+    }
+
+    // Then try parsing as YAML
+    if let Ok(_) = serde_yaml::from_str::<serde_yaml::Value>(&content) {
+        debug!("File detected as YAML: {}", file_path.display());
+        return Ok(InputFormat::Yaml);
+    }
+
+    // If neither worked, return an error
+    bail!("File is neither valid JSON nor valid YAML: {}", file_path.display())
+}
+
+/// Converts input file to JSON if necessary.
+///
+/// If the file is already JSON, returns the original path.
+/// If the file is YAML, converts it to JSON and returns the path to the temporary JSON file.
+pub fn get_json_path(input_path: &PathBuf) -> Result<PathBuf> {
```

Comment:
Instead of taking a reference and potentially cloning the input path, this function should take ownership of `input_path` and return it directly